pkg/manager/internal.go

```diff
@@ -161,57 +168,27 @@ type controllerManager struct {
 	// retryPeriod is the duration the LeaderElector clients should wait
 	// between tries of actions.
 	retryPeriod time.Duration
-}
-
-type errSignaler struct {
-	// errSignal indicates that an error occurred, when closed.  It shouldn't
-	// be written to.
-	errSignal chan struct{}
-
-	// err is the received error
-	err error
-
-	mu sync.Mutex
-}
-
-func (r *errSignaler) SignalError(err error) {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-
-	if err == nil {
-		// non-error, ignore
-		log.Error(nil, "SignalError called without an (with a nil) error, which should never happen, ignoring")
-		return
-	}
-
-	if r.err != nil {
-		// we already have an error, don't try again
-		return
-	}
 
-	// save the error and report it
-	r.err = err
-	close(r.errSignal)
-}
+	// waitForRunnable is holding the number of runnables currently running so that
+	// we can wait for them to exit before quitting the manager
+	waitForRunnable sync.WaitGroup
 
-func (r *errSignaler) Error() error {
-	r.mu.Lock()
-	defer r.mu.Unlock()
+	// gracefulShutdownTimeout is the duration given to runnable to stop
+	// before the manager actually returns on stop.
+	gracefulShutdownTimeout time.Duration
```

Comment:
alternatively, we should have a good way to detect dealine exceeded errors, the same was context does

Reply:
> is there a way to say "no timeout"?

I don't think so, except for maybe setting a multi year timeout

> alternatively, we should have a good way to detect dealine exceeded errors, the same was context does

I have changed the shutdown to use a `context` rather than a timer and propagate its error back + extended the tests to verify that we can find it via `errors.Is`

Reply:
time.Duration supports negative numbers.  You could detect a negative and use if/else to set the shutdownCtx appropriately.

Reply:
Great idea, done